File: modules/services/share-video/src/utils/transcribers/speechkit.ts
import * as fs from 'fs';
import * as crypto from 'crypto';
import { DeleteObjectCommand, GetObjectCommand, PutObjectCommand, S3Client } from '@aws-sdk/client-s3';
import { getSignedUrl } from '@aws-sdk/s3-request-presigner';
import { Transcriber, TranscribeOptions, TranscriptionResult, WordTimestamp } from './types';

const RECOGNIZE_URL = 'https://stt.api.cloud.yandex.net/stt/v3/recognizeFileAsync';
const OPERATIONS_URL = 'https://operation.api.cloud.yandex.net/operations';

export interface SpeechKitOptions {
  apiKey: string;
  s3: S3Client;
  bucket: string;
  /** Where the temp audio upload lives. Function must have RW on this prefix. */
  scratchPrefix: string;
  /** Where the bucket actually lives — used to build the audio source URI. */
  storageEndpoint?: string;
  /** Hard cap on poll wait. */
  pollTimeoutMs?: number;
  pollIntervalMs?: number;
}

/**
 * Yandex SpeechKit v3 async file recognition with word-level timestamps.
 *
 * Flow:
 *   1. PUT cut audio into the scratch prefix on the function's bucket.
 *   2. POST to recognizeFileAsync with the audio's https:// URI on YC
 *      Object Storage — SpeechKit pulls the file itself.
 *   3. Poll the operation until done (typically 5-30s for our 60-120s clips).
 *   4. Flatten chunks/alternatives into a flat word stream.
 *   5. DELETE the scratch upload (best-effort; not fatal if it fails).
 */
export function createSpeechKitTranscriber(opts: SpeechKitOptions): Transcriber {
  const pollTimeoutMs = opts.pollTimeoutMs ?? 5 * 60_000;
  const pollIntervalMs = opts.pollIntervalMs ?? 2_000;

  return {
    async transcribe(audioPath: string, options: TranscribeOptions): Promise<TranscriptionResult> {
      if (!fs.existsSync(audioPath)) throw new Error(`audio file not found: ${audioPath}`);

      const scratchKey = `${opts.scratchPrefix.replace(/\/$/, '')}/${crypto.randomBytes(16).toString('hex')}.mp3`;

      // 1. Upload the cut audio to YC Object Storage.
      await opts.s3.send(
        new PutObjectCommand({
          Bucket: opts.bucket,
          Key: scratchKey,
          Body: fs.readFileSync(audioPath),
          ContentType: 'audio/mpeg',
        }),
      );

      // Signed URL so SpeechKit can fetch the scratch file regardless of bucket policy.
      const audioUri = await getSignedUrl(
        opts.s3,
        new GetObjectCommand({ Bucket: opts.bucket, Key: scratchKey }),
        { expiresIn: 60 * 10 },
      );

      try {
        // 2. Start async recognition.
        const startBody = {
          uri: audioUri,
          recognitionModel: {
            model: 'general',
            audioFormat: { containerAudio: { containerAudioType: 'MP3' } },
            textNormalization: {
              textNormalization: 'TEXT_NORMALIZATION_ENABLED',
              profanityFilter: false,
              literatureText: false,
            },
            languageRestriction: {
              restrictionType: 'WHITELIST',
              languageCode: [toYcLocale(options.language)],
            },
          },
        };
        const startRes = await fetch(RECOGNIZE_URL, {
          method: 'POST',
          headers: {
            Authorization: `Api-Key ${opts.apiKey}`,
            'Content-Type': 'application/json',
          },
          body: JSON.stringify(startBody),
        });
        if (!startRes.ok) {
          throw new Error(`speechkit start ${startRes.status}: ${await startRes.text()}`);
        }
        const startJson: any = await startRes.json();
        const operationId = startJson.id;
        if (!operationId) {
          throw new Error(`speechkit: no operation id in response: ${JSON.stringify(startJson)}`);
        }

        // 3. Poll the operation.
        const deadline = Date.now() + pollTimeoutMs;
        let done: any = null;
        while (Date.now() < deadline) {
          await new Promise((r) => setTimeout(r, pollIntervalMs));
          const r = await fetch(`${OPERATIONS_URL}/${operationId}`, {
            headers: { Authorization: `Api-Key ${opts.apiKey}` },
          });
          if (!r.ok) throw new Error(`speechkit poll ${r.status}: ${await r.text()}`);
          const j: any = await r.json();
          if (j.done) {
            done = j;
            break;
          }
        }
        if (!done) throw new Error('speechkit: operation timed out');
        if (done.error) {
          throw new Error(`speechkit error: ${JSON.stringify(done.error)}`);
        }

        // 4. Parse result. v3 file-async returns { response: { chunks: [...] } }.
        return parseFileAsyncResult(done.response);
      } finally {
        // 5. Best-effort scratch cleanup.
        try {
          await opts.s3.send(new DeleteObjectCommand({ Bucket: opts.bucket, Key: scratchKey }));
        } catch (e: any) {
          console.warn(`speechkit scratch cleanup failed for ${scratchKey}: ${e?.message}`);
        }
      }
    },
  };
}

function parseFileAsyncResult(response: any): TranscriptionResult {
  // SpeechKit v3 file-async response shape:
  //   { chunks: [{ alternatives: [{ text, words: [{word, startTime, endTime}], ... }], channelTag }] }
  // OR streaming-result shape with finals — be defensive.
  const chunks: any[] = response?.chunks || response?.results || [];
  const text: string[] = [];
  const words: WordTimestamp[] = [];
  for (const chunk of chunks) {
    const alt = chunk?.alternatives?.[0] ?? chunk?.final?.alternatives?.[0];
    if (!alt) continue;
    if (alt.text) text.push(String(alt.text));
    for (const w of alt.words ?? []) {
      const word = String(w.word ?? w.text ?? '').trim();
      if (!word) continue;
      const start = parseDurationToSec(w.startTime ?? w.start_time ?? w.startTimeMs ?? w.start);
      const end = parseDurationToSec(w.endTime ?? w.end_time ?? w.endTimeMs ?? w.end);
      words.push({ word, start, end });
    }
  }
  return { text: text.join(' ').replace(/\s+/g, ' ').trim(), words };
}

/**
 * SpeechKit duration values are Google protobuf strings like "12.345s" — parse to seconds.
 * Tolerates plain numbers (already seconds) and ms-suffixed forms.
 */
function parseDurationToSec(v: unknown): number {
  if (v == null) return 0;
  if (typeof v === 'number') return v;
  const s = String(v).trim();
  let m = s.match(/^(-?\d+(?:\.\d+)?)s$/);
  if (m) return parseFloat(m[1]);
  m = s.match(/^(-?\d+(?:\.\d+)?)ms$/);
  if (m) return parseFloat(m[1]) / 1000;
  const n = Number(s);
  return Number.isFinite(n) ? n : 0;
}

const COMMON_LOCALE: Record<string, string> = {
  ru: 'ru-RU',
  en: 'en-US',
  uk: 'uk-UA',
  kk: 'kk-KK',
  tr: 'tr-TR',
  uz: 'uz-UZ',
  de: 'de-DE',
};

function toYcLocale(iso639: string): string {
  return COMMON_LOCALE[iso639.toLowerCase()] ?? iso639;
}

File: modules/services/share-video/internal/pipeline/transcript/speechkit.go
package transcript

import (
	"bytes"
	"context"
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"strconv"
	"strings"
	"time"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/s3"
)

const (
	speechKitRecognizeURL  = "https://stt.api.cloud.yandex.net/stt/v3/recognizeFileAsync"
	speechKitOperationsURL = "https://operation.api.cloud.yandex.net/operations"
	speechKitPollTimeout   = 5 * time.Minute
	speechKitPollInterval  = 2 * time.Second
)

type speechKit struct {
	apiKey      string
	s3c         *s3.Client
	presigner   *s3.PresignClient
	bucket      string
	scratchPref string
	httpc       *http.Client
}

func newSpeechKit(d Deps) *speechKit {
	return &speechKit{
		apiKey:      d.SpeechKitKey,
		s3c:         d.S3,
		presigner:   d.S3Presigner,
		bucket:      d.Bucket,
		scratchPref: strings.TrimRight(d.ScratchPrefix, "/"),
		httpc: &http.Client{
			Timeout: 5 * time.Minute,
		},
	}
}

type speechKitStartReq struct {
	URI              string `json:"uri"`
	RecognitionModel struct {
		Model       string `json:"model"`
		AudioFormat struct {
			ContainerAudio struct {
				ContainerAudioType string `json:"containerAudioType"`
			} `json:"containerAudio"`
		} `json:"audioFormat"`
		TextNormalization struct {
			TextNormalization string `json:"textNormalization"`
			ProfanityFilter   bool   `json:"profanityFilter"`
			LiteratureText    bool   `json:"literatureText"`
		} `json:"textNormalization"`
		LanguageRestriction struct {
			RestrictionType string   `json:"restrictionType"`
			LanguageCode    []string `json:"languageCode"`
		} `json:"languageRestriction"`
	} `json:"recognitionModel"`
}

type speechKitOpResp struct {
	ID       string          `json:"id"`
	Done     bool            `json:"done"`
	Error    json.RawMessage `json:"error,omitempty"`
	Response json.RawMessage `json:"response,omitempty"`
}

type speechKitChunk struct {
	Alternatives []struct {
		Text  string `json:"text"`
		Words []struct {
			Word      string `json:"word"`
			StartTime string `json:"startTime"`
			EndTime   string `json:"endTime"`
		} `json:"words"`
	} `json:"alternatives"`
}

type speechKitFileResp struct {
	Chunks []speechKitChunk `json:"chunks"`
}

func (sk *speechKit) Transcribe(ctx context.Context, audioPath string, opts Options) (Result, error) {
	body, err := os.ReadFile(audioPath)
	if err != nil {
		return Result{}, fmt.Errorf("read %s: %w", audioPath, err)
	}

	scratchKey := fmt.Sprintf("%s/%s.mp3", sk.scratchPref, randHex16())

	if _, err := sk.s3c.PutObject(ctx, &s3.PutObjectInput{
		Bucket:      aws.String(sk.bucket),
		Key:         aws.String(scratchKey),
		Body:        bytes.NewReader(body),
		ContentType: aws.String("audio/mpeg"),
	}); err != nil {
		return Result{}, fmt.Errorf("upload scratch: %w", err)
	}
	defer func() {
		// Best-effort cleanup; don't fail the whole transcription on this.
		_, _ = sk.s3c.DeleteObject(context.Background(), &s3.DeleteObjectInput{
			Bucket: aws.String(sk.bucket),
			Key:    aws.String(scratchKey),
		})
	}()

	presign, err := sk.presigner.PresignGetObject(ctx, &s3.GetObjectInput{
		Bucket: aws.String(sk.bucket),
		Key:    aws.String(scratchKey),
	}, s3.WithPresignExpires(10*time.Minute))
	if err != nil {
		return Result{}, fmt.Errorf("presign: %w", err)
	}

	var startReq speechKitStartReq
	startReq.URI = presign.URL
	startReq.RecognitionModel.Model = "general"
	startReq.RecognitionModel.AudioFormat.ContainerAudio.ContainerAudioType = "MP3"
	startReq.RecognitionModel.TextNormalization.TextNormalization = "TEXT_NORMALIZATION_ENABLED"
	startReq.RecognitionModel.LanguageRestriction.RestrictionType = "WHITELIST"
	startReq.RecognitionModel.LanguageRestriction.LanguageCode = []string{toYCLocale(opts.Language)}

	reqBody, _ := json.Marshal(startReq)
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, speechKitRecognizeURL, bytes.NewReader(reqBody))
	if err != nil {
		return Result{}, fmt.Errorf("build start request: %w", err)
	}
	req.Header.Set("Authorization", "Api-Key "+sk.apiKey)
	req.Header.Set("Content-Type", "application/json")
	resp, err := sk.httpc.Do(req)
	if err != nil {
		return Result{}, fmt.Errorf("speechkit start: %w", err)
	}
	startBody, _ := io.ReadAll(io.LimitReader(resp.Body, 1<<20))
	resp.Body.Close()
	if resp.StatusCode/100 != 2 {
		return Result{}, fmt.Errorf("speechkit start http %d: %s", resp.StatusCode, string(startBody))
	}

	var op speechKitOpResp
	if err := json.Unmarshal(startBody, &op); err != nil {
		return Result{}, fmt.Errorf("decode start response: %w", err)
	}
	if op.ID == "" {
		return Result{}, fmt.Errorf("speechkit: no operation id in response: %s", string(startBody))
	}

	deadline := time.Now().Add(speechKitPollTimeout)
	for time.Now().Before(deadline) {
		select {
		case <-ctx.Done():
			return Result{}, ctx.Err()
		case <-time.After(speechKitPollInterval):
		}
		preq, err := http.NewRequestWithContext(ctx, http.MethodGet, speechKitOperationsURL+"/"+op.ID, nil)
		if err != nil {
			return Result{}, fmt.Errorf("build poll request: %w", err)
		}
		preq.Header.Set("Authorization", "Api-Key "+sk.apiKey)
		presp, err := sk.httpc.Do(preq)
		if err != nil {
			return Result{}, fmt.Errorf("speechkit poll: %w", err)
		}
		pbody, _ := io.ReadAll(io.LimitReader(presp.Body, 16<<20))
		presp.Body.Close()
		if presp.StatusCode/100 != 2 {
			return Result{}, fmt.Errorf("speechkit poll http %d: %s", presp.StatusCode, string(pbody))
		}
		var pop speechKitOpResp
		if err := json.Unmarshal(pbody, &pop); err != nil {
			return Result{}, fmt.Errorf("decode poll response: %w", err)
		}
		if pop.Done {
			if len(pop.Error) > 0 && string(pop.Error) != "null" {
				return Result{}, fmt.Errorf("speechkit error: %s", string(pop.Error))
			}
			return parseSpeechKitFileResp(pop.Response)
		}
	}
	return Result{}, fmt.Errorf("speechkit: operation timed out after %s", speechKitPollTimeout)
}

func parseSpeechKitFileResp(raw json.RawMessage) (Result, error) {
	if len(raw) == 0 || string(raw) == "null" {
		return Result{}, fmt.Errorf("speechkit: empty response")
	}
	var resp speechKitFileResp
	if err := json.Unmarshal(raw, &resp); err != nil {
		return Result{}, fmt.Errorf("decode file response: %w", err)
	}
	var texts []string
	var words []Word
	for _, chunk := range resp.Chunks {
		if len(chunk.Alternatives) == 0 {
			continue
		}
		alt := chunk.Alternatives[0]
		if alt.Text != "" {
			texts = append(texts, alt.Text)
		}
		for _, w := range alt.Words {
			ws := strings.TrimSpace(w.Word)
			if ws == "" {
				continue
			}
			words = append(words, Word{
				Word:  ws,
				Start: parseDurationToSec(w.StartTime),
				End:   parseDurationToSec(w.EndTime),
			})
		}
	}
	return Result{
		Text:  collapseSpaces(strings.Join(texts, " ")),
		Words: words,
	}, nil
}

// parseDurationToSec accepts protobuf duration strings ("12.345s",
// "678ms") as well as plain numbers; returns seconds.
func parseDurationToSec(s string) float64 {
	s = strings.TrimSpace(s)
	if s == "" {
		return 0
	}
	if strings.HasSuffix(s, "ms") {
		n, _ := strconv.ParseFloat(strings.TrimSuffix(s, "ms"), 64)
		return n / 1000
	}
	if strings.HasSuffix(s, "s") {
		n, _ := strconv.ParseFloat(strings.TrimSuffix(s, "s"), 64)
		return n
	}
	n, _ := strconv.ParseFloat(s, 64)
	return n
}

func collapseSpaces(s string) string {
	out := strings.Builder{}
	prevSpace := false
	for _, r := range s {
		if r == ' ' || r == '\t' || r == '\n' || r == '\r' {
			if prevSpace {
				continue
			}
			out.WriteByte(' ')
			prevSpace = true
		} else {
			out.WriteRune(r)
			prevSpace = false
		}
	}
	return strings.TrimSpace(out.String())
}

var ycLocale = map[string]string{
	"ru": "ru-RU",
	"en": "en-US",
	"uk": "uk-UA",
	"kk": "kk-KK",
	"tr": "tr-TR",
	"uz": "uz-UZ",
	"de": "de-DE",
}

func toYCLocale(iso string) string {
	if v, ok := ycLocale[strings.ToLower(iso)]; ok {
		return v
	}
	return iso
}

func randHex16() string {
	var b [16]byte
	_, _ = rand.Read(b[:])
	return hex.EncodeToString(b[:])
}
